server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -362,24 +367,32 @@ public ClusterState execute(ClusterState currentState) {
                     // Restore templates (but do NOT overwrite existing templates)
                     restoreTemplates(mdBuilder, currentState);
 
-                    // Restore global state if needed
-                    if (request.includeGlobalState()) {
-                        if (metadata.persistentSettings() != null) {
-                            Settings settings = metadata.persistentSettings();
-                            clusterSettings.validateUpdate(settings);
-                            mdBuilder.persistentSettings(settings);
-                        }
-                        if (metadata.templates() != null) {
-                            // TODO: Should all existing templates be deleted first?
-                            for (ObjectCursor<IndexTemplateMetadata> cursor : metadata.templates().values()) {
-                                mdBuilder.put(cursor.value);
+                    if (request.includeGlobalSettings() && metadata.persistentSettings() != null) {
+                        Settings settings = metadata.persistentSettings();
+
+                        // CrateDB patch to only restore defined settings
+                        if (request.globalSettings().length > 0) {
+                            var filteredSettingBuilder = Settings.builder();
+                            for (String prefix : request.globalSettings()) {
+                                filteredSettingBuilder.put(settings.filter(s -> s.startsWith(prefix)));
                             }
+                            settings = filteredSettingBuilder.build();
                         }
-                        if (metadata.customs() != null) {
-                            for (ObjectObjectCursor<String, Metadata.Custom> cursor : metadata.customs()) {
-                                if (!RepositoriesMetadata.TYPE.equals(cursor.key)) {
-                                    // Don't restore repositories while we are working with them
-                                    // TODO: Should we restore them at the end?
+
+                        clusterSettings.validateUpdate(settings);
+                        mdBuilder.persistentSettings(settings);
+                    }
+                    if (request.includeCustomMetadata() && metadata.customs() != null) {
+                        // CrateDB patch to only restore defined custom metadata types
+                        List<String> customMetadataTypes = Arrays.asList(request.customMetadataTypes());
+                        boolean includeAll = customMetadataTypes.size() == 0;
+
+                        for (ObjectObjectCursor<String, Metadata.Custom> cursor : metadata.customs()) {
+                            if (!RepositoriesMetadata.TYPE.equals(cursor.key)) {
+                                // Don't restore repositories while we are working with them
+                                // TODO: Should we restore them at the end?
```

Comment:
What would happen if we do restore them?

I assume it could delete a repository that's actively being used? What would happen in that case?

Reply:
Good point, didn't took care of it in general as this was a bit unrelated to my changes.
I think this would require a bit more of testing. I suggest to followup here dedicated.